### PR TITLE
[SC-105] Rewrite tests & remove 'aave-helpers' dependency

### PR DESCRIPTION
**Pull request by barrutko**

### CODE REVIEW

File: foundry.toml [hunk 1]
```diff
@@ -3,9 +3,7 @@ src = "src"
 out = "out"
 libs = ["lib"]
 remappings = [
-    'aave-helpers/=lib/aave-helpers/src/',
-    'solidity-utils/=lib/aave-helpers/lib/solidity-utils/src/',
-    'aave-address-book/=lib/aave-helpers/lib/aave-address-book/src'
+    'xchain-helpers/=lib/xchain-helpers/src/'
```
Comment (hexonaut) [inline foundry.toml]: This should be picked up automatically without needing to manually specify the remapping.

File: .gitmodules [hunk 2]
```diff
@@ -1,6 +1,7 @@
 [submodule "lib/forge-std"]
 	path = lib/forge-std
 	url = https://github.com/foundry-rs/forge-std
-[submodule "lib/aave-helpers"]
-	path = lib/aave-helpers
-	url = https://github.com/marsfoundation/aave-helpers
+[submodule "lib/xchain-helpers"]
+	path = lib/xchain-helpers
+	url = https://github.com/marsfoundation/xchain-helpers
+	branch = SC-97-init
```
Comment (hexonaut) [inline .gitmodules]: I think we can leave the branch off now as pending your approval that will be merged into master.

**Review by hexonaut — COMMENTED** [review]
Two minor issues. This is a good first pass translating the previous version. I'd like to do another PR after where the test logic is merged into a single base test contract and the specifics of the domain are dealt with by inheriting contracts. Similar to how this is done: https://github.com/makerdao/dss-bridge/tree/v1/src/tests/domains

Would also like to see the testing expanded to cover more of the features, but I think ordering can be so as to merge the test contracts first then adding more tests after.

File: .gitmodules (hunk 2 above)
Comment (barrutko) [reply .gitmodules]: Updated

**Comment by barrutko** [comment]
Redone the structure and working on tests. Still some more to go.

**Review by hexonaut — CHANGES_REQUESTED** [review]
The `xchain-helpers` appears to be referencing a non-existent commit. Not sure what's going on here.

**Comment by barrutko** [comment]
> The `xchain-helpers` appears to be referencing a non-existent commit. Not sure what's going on here.

Should be fine now.

**Review by hexonaut — APPROVED** [review]
LGTM

File: test/CrosschainTestBase.sol [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    uint256 maximumDelay;
+    address guardian;
+}
+
+abstract contract CrosschainTestBase is Test  {
+    event TestEvent();
+
+    address public constant L1_EXECUTOR    = 0x3300f198988e4C9C63F75dF86De36421f06af8c4;
+    address public constant L1_PAUSE_PROXY = 0xBE8E3e3618f7474F8cB1d074A26afFef007E98FB;
+    address public constant GUARDIAN       = 0x474E6f886fE829Fd5F289C5B681DdE09ab207076;
```
Comment (hexonaut) [inline test/CrosschainTestBase.sol]: What is this address? If you need a valid, random address it's better to use `makeAddr("guardian")` or similar.

File: test/CrosschainTestBase.sol [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+import { Domain } from 'xchain-helpers/Domain.sol';
+import { BridgedDomain } from 'xchain-helpers/BridgedDomain.sol';
+
+import { IL2BridgeExecutor, IExecutorBase } from '../src/interfaces/IL2BridgeExecutor.sol';
+
+import { PayloadWithEmit } from './mocks/PayloadWithEmit.sol';
+import { ReconfigurationPayload } from './mocks/ReconfigurationPayload.sol';
+import { IExecutor } from './interfaces/IExecutor.sol';
+
+interface IBaseCrossschainForwarder {
```
Comment (hexonaut) [inline test/CrosschainTestBase.sol]: Typo, 3 "s" in Crosss

File: test/CrosschainTestBase.sol [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            forwarder,
+            encodedPayloadData
+        );
+
+        bridgedDomain.relayFromHost(true);
+    }
+
+    function testFuzz_basicCrosschainPayloadExecution(uint256 delay) public {
+        vm.assume(delay > defaultL2BridgeExecutorArgs.delay);
+        vm.assume(delay < (defaultL2BridgeExecutorArgs.delay + defaultL2BridgeExecutorArgs.gracePeriod));
```
Comment (hexonaut) [inline test/CrosschainTestBase.sol]: Should not use assume when trying to convert a wide range into a small one as this `assume` will do trial and error to find valid values which is very wasteful. You should instead `bound()` function.

File: test/CrosschainTestBase.sol [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        skip(delay);
+
+        vm.expectEmit(bridgeExecutor);
+        emit TestEvent();
+        IL2BridgeExecutor(bridgeExecutor).execute(0);
+    }
+
+    function testFuzz_actionExecutionFailsAfterGracePeriod(uint delay) public {
+        vm.assume(delay < 10_000_000);
+        vm.assume(delay > (defaultL2BridgeExecutorArgs.delay + defaultL2BridgeExecutorArgs.gracePeriod));
```
Comment (hexonaut) [inline test/CrosschainTestBase.sol]: Same here.

File: test/CrosschainTestBase.sol [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        preparePayloadExecution();
+
+        skip(delay);
+
+        vm.expectRevert(IExecutorBase.OnlyQueuedActions.selector);
+        IL2BridgeExecutor(bridgeExecutor).execute(0);
+    }
+
+    function testFuzz_actionxecutionFailsBeforeTimelock(uint delay) public {
+        vm.assume(delay < defaultL2BridgeExecutorArgs.delay);
```
Comment (hexonaut) [inline test/CrosschainTestBase.sol]: Same here.

File: test/CrosschainTestBase.sol [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        vm.expectRevert();
+        IL2BridgeExecutor(bridgeExecutor).execute(1);
+
+        vm.expectEmit(bridgeExecutor);
+        emit TestEvent();
+        IL2BridgeExecutor(bridgeExecutor).execute(0);
+    }
+
+    function test_onlyGuardianCanCancel() public {
+        address notGuardian = 0x17B23Be942458E6EfC17F000976A490EC428f49A;
```
Comment (hexonaut) [inline test/CrosschainTestBase.sol]: Again I would use `makeAddr("notGuardian")`.

File: test/CrosschainTestBase.sol [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+        assertEq(
+            IL2BridgeExecutor(bridgeExecutor).getMaximumDelay(),
+            defaultL2BridgeExecutorArgs.maximumDelay
+        );
+        assertEq(
+            IL2BridgeExecutor(bridgeExecutor).getGuardian(),
+            defaultL2BridgeExecutorArgs.guardian
+        );
+
+        bridgedDomain.selectFork();
```
Comment (hexonaut) [inline test/CrosschainTestBase.sol]: You are already on the L2 fork.

File: test/mocks/ReconfigurationPayload.sol [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    function getNewMinimumDelay() public pure returns (uint256) {
+        return 100;
+    }
+
+    function getNewMaximumDelay() public pure returns (uint256) {
+        return 3600;
+    }
+
+    function getNewGuardian() public pure returns (address) {
+        return 0x3300f198988e4C9C63F75dF86De36421f06af8c4;
```
Comment (hexonaut) [inline test/mocks/ReconfigurationPayload.sol]: Would pass in a `makeAddr("newGuardian")` in the constructor.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Sorry I jumped the gun a bit with my last review. I didn't realize you merged 3 tasks into this PR.

Overall looks good, just some best practises to follow.

File: test/mocks/ReconfigurationPayload.sol (hunk 10 above)
Comment (barrutko) [reply test/mocks/ReconfigurationPayload.sol]: I can't really do that. It's gonna be in the payload storage and the delegate call from executor is gonna run on the executor's storage.

Comment (barrutko) [reply test/mocks/ReconfigurationPayload.sol]: For now changed it to something more obviously meaningless.

Comment (barrutko) [reply test/mocks/ReconfigurationPayload.sol]: Ok, immutable is the thing 😅 Fixed.

**Review by hexonaut — DISMISSED** [review]
LGTM

File: test/CrosschainTestBase.sol [hunk 11]
```diff
@@ -249,7 +249,22 @@ abstract contract CrosschainTestBase is Test  {
...
 
-        ReconfigurationPayload reconfigurationPayload = new ReconfigurationPayload(makeAddr("newGuardian"));
+        L2BridgeExecutorArguments memory newL2BridgeExecutorParams = L2BridgeExecutorArguments({
+                ethereumGovernanceExecutor: defaultL2BridgeExecutorArgs.ethereumGovernanceExecutor,
+                delay:                      1200,
+                gracePeriod:                1800,
+                minimumDelay:               100,
+                maximumDelay:               3600,
+                guardian:                   makeAddr("newGuardian")
+            });
```
Comment (hexonaut) [inline test/CrosschainTestBase.sol]: Tab alignment.

**Review by hexonaut — CHANGES_REQUESTED** [review]

File: test/CrosschainTestBase.sol (hunk 11 above)
Comment (barrutko) [reply test/CrosschainTestBase.sol]: Fixed

**Review by hexonaut — DISMISSED** [review]

File: test/ArbitrumCrosschainTest.t.sol [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        hostDomain.selectFork();
+        forwarder = address(new CrosschainForwarderArbitrum(bridgeExecutor));
+        vm.deal(
+            L1_EXECUTOR,
+            0.01 ether
+        );
+    }
+
+    function test_arbitrumGasCalculations(uint256 paylodDataLength) public {
+        paylodDataLength = bound(
```
Comment (barrutko) [inline test/ArbitrumCrosschainTest.t.sol]:
```suggestion
        payloadDataLength = bound(
```

File: test/ArbitrumCrosschainTest.t.sol [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+        hostDomain.selectFork();
+        forwarder = address(new CrosschainForwarderArbitrum(bridgeExecutor));
+        vm.deal(
+            L1_EXECUTOR,
+            0.01 ether
+        );
+    }
+
+    function test_arbitrumGasCalculations(uint256 paylodDataLength) public {
```
Comment (barrutko) [inline test/ArbitrumCrosschainTest.t.sol]:
```suggestion
    function testFuzz_arbitrumGasCalculations(uint256 paylodDataLength) public {
```

File: test/CrosschainTestBase.sol [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+
+        preparePayloadExecution();
+
+        skip(delay);
+
+        vm.expectRevert(IExecutorBase.OnlyQueuedActions.selector);
+        IL2BridgeExecutor(bridgeExecutor).execute(0);
+    }
+
+    function testFuzz_actionxecutionFailsBeforeTimelock(uint delay) public {
```
Comment (barrutko) [inline test/CrosschainTestBase.sol]:
```suggestion
    function testFuzz_actionExecutionFailsBeforeTimelock(uint delay) public {
```

**Review by lucas-manuel — APPROVED** [review]
LGTM for a start, will build on this and add more testing as needed

**Review by hexonaut — APPROVED** [review]

**Pull request merged**